Source/DafnyRuntime/DafnyRuntime.go

```diff
@@ -692,183 +692,412 @@ func (seq Seq) VerbatimString(asLiteral bool) string {
  * Arrays
  ******************************************************************************/
 
-// An Array is a Go slice representing a (possibly) multidimensional array,
-// along with metadata.  There aren't any methods for updating; instead, you can
-// update through the pointer returned by Index.
-type Array struct {
-  contents []interface{} // stored as a flat one-dimensional slice
-  dims     []int
-}
-
-func newArray(dims ...Int) *Array {
+// Array is the general interface for arrays. Conceptually, it contains some
+// underlying storage (a slice) of elements, together with a record of the length
+// of each dimension. Thus, this general interface supports 1- and multi-dimensional
+// Dafny arrays.
+//
+// All array indices are given as the Go type "int". This is what the Dafny implementation
+// refers to as "target-language array-index type". However, the dimension lengths
+// are given as "big.Int".
+
+type Array interface {
+  dimensionCount() int
+  dimensionLength(dim int) int
+  ArrayGet1(index int) interface{}
+  ArraySet1(value interface{}, index int)
+  anySlice(lo, hi Int) []interface{}
+  // specializations
+  ArrayGet1Byte(index int) byte
+  ArraySet1Byte(value byte, index int)
+  ArrayGet1Char(index int) Char
+  ArraySet1Char(value Char, index int)
+}
+
+/***** newArray *****/
+
+// Multiply the numbers in "dims" and return the product as an "int".
+// If the produce doesn't fit in an "int", panic with the message that the
+// array-size limit has been exceeded.
+// It is expected that len(dims) is at least 1 and that each number in
+// dims is non-negative.
+func computeTotalArrayLength(dims ...Int) int {
+  product := dims[0]
+  for i := 1; i < len(dims); i++ {
+    product = product.Times(dims[i])
+  }
+  if IntOf(0x8000_0000).Cmp(product) <= 0 {
+    panic(fmt.Sprintf("array size exceeds memory limit: %v", String(product)))
+  }
+  totalLength := product.Int()
+  return totalLength
+}
+
+// NewArrayFromExample returns a new Array.
+// If "init" is non-nil, it is used to initialize all elements of the array.
+// "example" is used only to figure out the right kind of Array to return.
+// If "init" is non-nil, the types of "example" and "init" must agree.
+func NewArrayFromExample(example interface{}, init interface{}, dims ...Int) Array {
```

Comment:
So I mentioned earlier this concern that determining the array element type based on an example value could make the wrong choice: if the Dafny type of the array is `array<T>` where `T` is a trait and `C1` is one class that extends it, and we provide a `C1` instance as the example, we could allocate a type-safe array of `C1`s that then later fails to accept an element of type `C2` (another class that extends `T`).

That may be a reasonable limitation of this technique and may not actually be triggerable YET, but could sneak in if we continue to apply this technique, and especially if the Dafny type system evolves such that we have more cases of subtyping that could trigger it (such as allowing value types to implement traits as well). Is this indeed safe now, and if so how do we ensure it remains safe in the future?

Reply:
This is a good point. It's not currently a problem, because for now in Go we just need to figure out the size of the array elements. Since trait `T` and class `C1` have the same size, we're good.

As you mention, we'll have to be careful about this in the future if we allow a byte/char/codepoint to implement a trait. My guess of how it will play out is: If we're creating an `array<T>` where `T` is a trait that is implemented by, say, a Dafny `byte` type, then the example element will be a fat pointer to a `T`. From this, we'll be able to tell, at run time, that the array-element size should be that of a pointer, not that of a byte. In other words, as long as the example element is really passed in using the representation we use for a trait, then the scheme in this PR will continue to work.

Reply:
That's fair, one possibility is to provide not just a `0x0` byte value but an augmented `(0x0, MyTrait)` value of some kind. Or perhaps even just a type descriptor value if that's simpler.

Reply:
I thought some more about this. From a correctness perspective, we should be fine. However, we may in the future want to optimize references to instances of _classes_. When such a reference is converted to an `interface{}`, I believe it goes from one machine word of storage to two. This is based on my knowledge of the _abstract_ representation of a interface reference in Go, which is a fat pointer. I don't know if the running Go actually stores such fat pointers as one or two machine words. Anyhow, if they are different and if we want to detect this at run time to (in some way I don't actually understand how) optimize the fat pointer to a single pointer, then the issue you're bringing up with traits is something we'll need to think about.